probe-rs/src/config/registry.rs

```diff
@@ -145,112 +145,109 @@ impl Registry {
+                    tracing::warn!(
+                        "Ignoring {matches} ambiguous matches for specified chip name {name}"
+                    );
+
+                    let (print, overflow) =
+                        partial_matches.split_at(MAX_PRINTED_MATCHES.min(matches));
+
+                    let mut suggestions = print.join(", ");
+
+                    match overflow.len() {
+                        0 => {}
+                        1 => suggestions.push_str(&format!(", {}", overflow[0])),
+                        _ => suggestions.push_str(&format!("and {} more", overflow.len())),
+                    }
+
+                    return Err(RegistryError::ChipNotUnique(name.to_string(), suggestions));
                 }
-                return Err(RegistryError::ChipNotUnique(name.to_owned(), suggestions));
-            }
-            let (family, chip) = selected_family_and_chip
-                .ok_or_else(|| RegistryError::ChipNotFound(name.to_owned()))?;
-            if exact_matches == 0 && partial_matches.len() == 1 {
-                tracing::warn!(
-                    "Found chip {} which matches given partial name {}. Consider specifying its full name.",
-                    chip.name,
-                    name,
-                );
-            }
-            if chip.name.to_ascii_lowercase() != name.to_ascii_lowercase() {
-                tracing::warn!(
-                    "Matching {} based on wildcard. Consider specifying the chip as {} instead.",
-                    name,
-                    chip.name,
-                );
             }
+        }
+
+        if !chip.name.eq_ignore_ascii_case(name) {
+            tracing::warn!(
+                "Matching {} based on wildcard. Consider specifying the chip as {} instead.",
+                name,
+                chip.name,
+            );
+        }
 
-            // Try get the correspnding flash algorithm.
-            (family, chip)
-        };
         let targ = self.get_target(family, chip)?;
         Ok((targ, family.clone()))
     }
 
-    fn get_targets_by_family_name(
-        &self,
-        family_name: impl AsRef<str>,
-    ) -> Result<Vec<String>, RegistryError> {
-        let name: &str = family_name.as_ref();
-
-        let family = {
-            let mut finded_family = None;
-            let mut exact_matches = 0;
-            for family in &self.families {
-                if match_name_prefix(&family.name, name) {
-                    if family.name.len() == name.len() {
-                        tracing::debug!("Exact match for family name: {}", family.name);
-                        exact_matches += 1;
-                    } else {
-                        tracing::debug!("Partial match for family name: {}", family.name);
-                        if exact_matches > 0 {
-                            continue;
-                        }
+    fn get_targets_by_family_name(&self, name: &str) -> Result<Vec<String>, RegistryError> {
```

Comment:
Why is &str preferable over AsRef<str>?

Reply:
This is just a tiny bit of best practice: this can avoid monomorphising a large function multiple times. Sine the ones I modified are private, they can only ever be called from generic wrappers that do the `AsRef -> str` conversion. As long as `get_targets_by_family_name` is not inlined, we end up with slightly less code generated.

https://www.possiblerust.com/pattern/non-generic-inner-functions

Reply:
Nice explanation! I knew this was the reason but I did not realize you do this in private functions :)